### PR TITLE
cluster groups proposal

### PR DESCRIPTION
**Pull request by deads2k**
Just cluster scoped groups, split from https://github.com/openshift/origin/pull/3627

@liggitt @kargakis @soltysh @stevekuznetsov

### CODE REVIEW

File: docs/proposals/cluster-groups.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+To allow cluster-scoped group definitions to be shared across multiple related projects, we'll introduce the concept of Groups (cluster-scoped sets of users) that are managed by cluster-admins.
+
+
+## Use Cases
+ 1.  I want to be able to define some Groups that will affect policy evaluation across all projects.
+ 1.  I want to be able to efficiently manage group membership.
+
+
+## Groups
+Groups are cluster-scoped sets of users that are managed by cluster-admins and may contain any user.  Groups can be bound to roles in all projects.  That means that if UserA is added to GroupB and ProjectC has a rolebinding that says "GroupB is an editor", UserA now has editor permissions in ProjectC.  Groups will be described using an API object that looks like this:
```
Comment (kargakis) [inline docs/proposals/cluster-groups.md]: s/managed/created and managed

Comment (soltysh) [inline docs/proposals/cluster-groups.md]: IIMHO managed consists of creating, modifying and deleting.

**Comment by soltysh** [comment]
LGTM

**Comment by deads2k** [comment]
@smarterclayton @pweil- any further comment?

**Comment by pweil-** [comment]
lgtm as well

File: docs/proposals/cluster-groups.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+type Group struct {
+	kapi.TypeMeta
+	kapi.ObjectMeta
+
+	// Users is a list of Users that are members of the Group
+	Users util.StringSet
+}
+```
+
+A reverse lookup index will be built in memory (similar to the project cache) to be allow the authenticator to make fast lookups for Group membership determination.
```
Comment (smarterclayton) [inline docs/proposals/cluster-groups.md]: What's the numbers on the size of the reverse index here?  If I have 1000 users who are each members of 50 groups in ldap, is the fan out manageable?  What about larger scales?

Comment (deads2k) [inline docs/proposals/cluster-groups.md]: > What's the numbers on the size of the reverse index here? If I have 1000 users who are each members of 50 groups in ldap, is the fan out manageable? What about larger scales?

For a while.
```
count(users) * len(username) * count(avg group memberships) * len(groups)
```

If we take your example: say `1024 * 32 * 64 * 32`, we come out with 64MB.  It's significant, but probably acceptable.

Comment (smarterclayton) [inline docs/proposals/cluster-groups.md]: Group names and usernames in ldap can get pretty long.  1000 users is low
end, but high end is going to be something like 20k

On Mon, Jul 27, 2015 at 2:46 PM, David Eads <notifications@github.com>
wrote:

> In docs/proposals/cluster-groups.md
> <https://github.com/openshift/origin/pull/3753#discussion_r35571403>:
>
> > +
> > +
> > +## Groups
> > +Groups are cluster-scoped sets of users that are managed by cluster-admins and may contain any user.  Groups can be bound to roles in all projects.  That means that if UserA is added to GroupB and ProjectC has a rolebinding that says "GroupB is an editor", UserA now has editor permissions in ProjectC.  Groups will be described using an API object that looks like this:
> > +```
> > +// Group contains the set of users that belong to a given Group.  The name may not contain ":"
> > +type Group struct {
> > +	kapi.TypeMeta
> > +	kapi.ObjectMeta
> > +
> > +	// Users is a list of Users that are members of the Group
> > +	Users util.StringSet
> > +}
> > +```
> > +
> > +A reverse lookup index will be built in memory (similar to the project cache) to be allow the authenticator to make fast lookups for Group membership determination.
>
>  What's the numbers on the size of the reverse index here? If I have 1000
> users who are each members of 50 groups in ldap, is the fan out manageable?
> What about larger scales?
>
> For a while.
>
> count(users) * len(username) * count(avg group memberships) * len(groups)
>
> If we take your example: say 1024 * 32 * 64 * 32, we come out with 64MB.
> It's significant, but probably acceptable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3753/files#r35571403>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

Comment (deads2k) [inline docs/proposals/cluster-groups.md]: > Group names and usernames in ldap can get pretty long.  1000 users is low
> end, but high end is going to be something like 20k

Some alternatives if/when we get there(all of them painful)
 1. tell people to be careful about which groups they sync
 1. allow users to specify an ldap query indicating which users to allow group syncing for (@stevekuznetsov potential future item)
 1. build a non-authoritative reverse list in a User for a lazy LRU cache
 1. build indexes into etcd.

Comment (smarterclayton) [inline docs/proposals/cluster-groups.md]: That's an acceptable answer.

On Mon, Jul 27, 2015 at 3:04 PM, David Eads <notifications@github.com>
wrote:

> In docs/proposals/cluster-groups.md
> <https://github.com/openshift/origin/pull/3753#discussion_r35573487>:
>
> > +
> > +
> > +## Groups
> > +Groups are cluster-scoped sets of users that are managed by cluster-admins and may contain any user.  Groups can be bound to roles in all projects.  That means that if UserA is added to GroupB and ProjectC has a rolebinding that says "GroupB is an editor", UserA now has editor permissions in ProjectC.  Groups will be described using an API object that looks like this:
> > +```
> > +// Group contains the set of users that belong to a given Group.  The name may not contain ":"
> > +type Group struct {
> > +	kapi.TypeMeta
> > +	kapi.ObjectMeta
> > +
> > +	// Users is a list of Users that are members of the Group
> > +	Users util.StringSet
> > +}
> > +```
> > +
> > +A reverse lookup index will be built in memory (similar to the project cache) to be allow the authenticator to make fast lookups for Group membership determination.
>
>  Group names and usernames in ldap can get pretty long. 1000 users is low
> end, but high end is going to be something like 20k
>
> Some alternatives if/when we get there(all of them painful)
>
>    1. tell people to be careful about which groups they sync
>    2. allow users to specify an ldap query indicating which users to
>    allow group syncing for (@stevekuznetsov
>    <https://github.com/stevekuznetsov> potential future item)
>    3. build a non-authoritative reverse list in a User for a lazy LRU
>    cache
>    4. build indexes into etcd.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3753/files#r35573487>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by deads2k** [comment]
[merge]

**Comment by deads2k** [comment]
merging before the impl

**Pull request merged**